codec/address/bech32_codec.go

```diff
@@ -3,23 +3,92 @@ package address
 import (
 	"errors"
 	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/hashicorp/golang-lru/simplelru"
 
 	"cosmossdk.io/core/address"
 	errorsmod "cosmossdk.io/errors"
 
+	"github.com/cosmos/cosmos-sdk/internal/conv"
 	sdkAddress "github.com/cosmos/cosmos-sdk/types/address"
 	"github.com/cosmos/cosmos-sdk/types/bech32"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
+const (
+	// TODO: ideally sdk.GetBech32PrefixValAddr("") should be used but currently there's a cyclical import.
+	// 	Once globals are deleted the cyclical import won't happen.
+	prefixValAddr  = "valoper"
+	prefixConsAddr = "valcons"
+)
+
+// cache variables
+var (
+	accAddrMu     sync.Mutex
+	accAddrCache  *simplelru.LRU
+	consAddrMu    sync.Mutex
+	consAddrCache *simplelru.LRU
+	valAddrMu     sync.Mutex
+	valAddrCache  *simplelru.LRU
+
+	isCachingEnabled atomic.Bool
+)
+
+func init() {
+	var err error
+	isCachingEnabled.Store(true)
```

Comment:
nit: this is always true

Reply:
Yeah, followed the same style as in [address.go](https://github.com/cosmos/cosmos-sdk/blob/8e60f3b7ebf4fb7b9d49a139ace62102d1af790f/types/address.go#L118-L120), although I think it doesn't make much sense.